src/Interpreters/InterpreterSelectQuery.h

```diff
@@ -29,6 +32,15 @@ class QueryPlan;
 
 struct TreeRewriterResult;
 using TreeRewriterResultPtr = std::shared_ptr<const TreeRewriterResult>;
+using Data = std::pair<Block, Chunks>;
+
+struct ASTHash
+{
+    size_t operator() (const IAST::Hash & hash) const
+    {
+        return hash.first * 7919 + hash.second;
```

Comment:
In cache we should put not only AST but also header type, because we will get logical errors.
Consider example
```sql
CREATE TABLE test_table (id UInt64) ENGINE=TinyLog;
SELECT id FROM test_table;
DROP TABLE test_table;
CREATE TABLE test_table (id String) ENGINE=TinyLog;
SELECT id FROM test_table;
```
this will lead to crash.

Reply:
Fixed.